agenta-backend/agenta_backend/services/evaluators_service.py

```diff
@@ -704,6 +708,61 @@ def auto_similarity_match(
         )
 
 
+async def semantic_similarity(output: str, correct_answer: str, api_key: str) -> float:
+    """Calculate the semantic similarity score of the LLM app using OpenAI's Embeddings API.
+
+    Args:
+        output (str): the output text
+        correct_answer (str): the correct answer text
+
+    Returns:
+        float: the semantic similarity score
+    """
+
+    openai = AsyncOpenAI(api_key=api_key)
+
+    async def encode(text: str):
+        response = await openai.embeddings.create(
+            model="text-embedding-3-small", input=text
+        )
+        return np.array(response.data[0].embedding)
+
+    def cosine_similarity(output_vector: array, correct_answer_vector: array) -> float:
+        return 1 - cosine(output_vector, correct_answer_vector)
+
+    output_vector = await encode(output)
+    correct_answer_vector = await encode(correct_answer)
+    return cosine_similarity(output_vector, correct_answer_vector)
```

Comment:
https://stackoverflow.com/questions/18424228/cosine-similarity-between-2-number-lists

`from numpy import dot`
`cos_sim = dot(a, b) # You can assume norm-1 embeddings from APIs, I think : needs double-check`

> Equivalent to: 1 - scipy.spatial.distance.cosine(a, b)
> Fewer dependencies
> Faster

`from scipy import spatial`
and
`from sklearn.metrics.pairwise import cosine_similarity`
> More dependencies
> Slower

Reply:
### ====== Benchmark ======

1).

| Datapoint              | Scipy | Numpy |
| :---------------- | :------: | ----: |
| 1        |   1.396359s   | 1.787759s |
| 2           |   1.203814s   | 1.405315s |
| 3    |  1.628027s   | 1.558809s |
| 4 |  1.951048s   | 1.570303s |
| 5 |  1.187300s   | 1.322176s|
| 6 |  1.207042s   | 1.298309s |
| 7 |  1.428319s   | 1.415392s |

![image](https://github.com/user-attachments/assets/9f948e25-6842-40f6-ba25-6c241d49064d)

2).

| Datapoint              | Scipy | Numpy |
| :---------------- | :------: | ----: |
| 1        |   1.405497s   | 1.309752s |
| 2           |   1.399423s   | 1.308726s |
| 3    |  1.374770s   | 1.421385s |
| 4 |  1.253854s   | 1.199923s |
| 5 |  1.549454s   | 1.640289s|
| 6 |  1.441418s  | 1.399796s |
| 7 |  1.152469s   | 1.415392s |

![image](https://github.com/user-attachments/assets/4a0e3c68-e63f-4268-9339-9ba0ece44df7)

Reply:
@aybruhm I think most of the time is spent on embedding rather than similarity calculation. The right benchmark would simply run cosine similarity on the two libraries 10,000 times and compute the average. Thanks to Claude, this could be done in less than a minute :)

Here are the results:
```
Benchmarking cosine similarity calculations for 10,000 pairs of vectors

Vector size: 1536

- NumPy dot: 0.0350 seconds
- SciPy spatial: 0.1908 seconds
```

Numpy is almost one order of magnitude faster. Although the bottleneck is clearly the embedding computation and not the similarity. 

@aybruhm I'd say let's go with numpy, it's faster (although it does not matter), and one less dependency.

Reply:
Oh, this is wonderful. Thank you for the results done well!